src/crystal/scheduler/runnables.cr

```diff
@@ -0,0 +1,146 @@
+      success
+    end
+  end
+
+  # :nodoc:
+  #
+  # Thread-safe non-blocking queue for work-stealing schedulers.
+  #
+  # Based on:
+  #
+  # - "Scheduling Multithreaded Computations by Work Stealing" (2001) by
+  #   Nimar S. Arora, Robert D. Blumofe and C. Greg Plaxton.
+  #
+  # - "Verification of a Concurrent Deque Implementation" (1999) by
+  #   Robert D. Blumofe, C. Greg Plaxton and Sandip Ray.
+  #
+  # The queue has the following assumptions:
+  #
+  # - Only the owner thread shall push to the bottom of the queue (never
+  #   call concurrently);
+  #
+  # - Only the owner thread shall pop from the bottom of the queue (never called
+  #   concurrently, and limited concurrency with pop top);
+  #
+  # - Only other threads shall pop from the top of the queue (expected to be
+  #   called concurrently):
+  #
+  # - The underlying array is supposed to be infinite. In practice, a scheduler
+  #   will only push/pop fibers from the bottom of the queue, which shouldn't
+  #   grow the queue much.
+  #
+  #   Thief schedulers will pop fibers from the top of the queue, digging a hole
+  #   at the top of the array, but it will be fixed whenever the queue is
+  #   emptied (bottom resets to 0).
+  #
+  #   This can however be an issue if a thread continuously pushes fibers to a
+  #   queue primarily meant to be stolen from (e.g. a dedicated event loop
+  #   thread with no scheduler).
+  #
+  # Refer to the papers above for algorithms, detailed analyzis and proofs.
+  class Runnables
+    # :nodoc:
+    record Age, tag : Int32, top : Int32
+
+    # :nodoc:
+    SIZE = Int32::MAX
+
+    def initialize
+      @bot = Atomic(Int32).new(0)
+      @age = AtomicRef8(Age).new(Age.new(0, 0))
+
+      # initializes the "infinite" array, using mmap so memory pages will only
+      # be allocated when accessed (if ever):
+      prot = LibC::PROT_READ | LibC::PROT_WRITE
+      flags = LibC::MAP_ANONYMOUS | LibC::MAP_PRIVATE
+      ptr = LibC.mmap(nil, SIZE, prot, flags, -1, 0)
+      raise Errno.new("mmap") if ptr == LibC::MAP_FAILED
+      @deq = ptr.as(Fiber*)
+    end
+
+    def free
+      LibC.munmap(@deq, SIZE)
+    end
+
+    # Pushes an item to the tail of the queue. Not thread-safe and must be
+    # called from the thread that owns the queue.
+    def push(item : Fiber) : Nil
+      bot = @bot.get
+      @deq[bot] = item
+      @bot.set(bot + 1)
+    end
+
+    # Pops an item from the tail of the queue. Not thread-safe and must be
+    # called from the thread that owns the queue.
+    def pop? : Fiber?
+      bot = @bot.get
+      return if bot == 0
+
+      bot -= 1
+      @bot.set(bot)
+
+      item = @deq[bot]
+      old_age = @age.get
+      return item if bot > old_age.top
+
+      # queue has been emptied (reset bottom to zero):
+      @bot.set(0)
+      new_age = Age.new(old_age.tag + 1, 0)
+
+      if bot == old_age.top
+        if @age.compare_and_set(old_age, new_age)
```

Comment:
AFAIK: `pop?` must win the CAS when popping the *last* fiber from the queue, because a parallel `shift?` may be trying to take it too. Either one that wins the CAS can return the fiber, the other one must return `nil`.